File: vol2/ch18/LeetCode1462-2.cpp
class Solution {
    vector<vector<int>> adj;            //邻接表
    vector<vector<bool>> path;                  //path[i][j]表示i到j是否有路径
public:
    vector<bool> checkIfPrerequisite(int n,vector<vector<int>>&prereqs,vector<vector<int>>&ques) {
        adj=vector<vector<int>>(n);
        path=vector<vector<bool>>(n,vector<bool>(n,false));
        for (int i=0;i<prereqs.size();i++) {    //由列表创建邻接表
            int a=prereqs[i][0];                //[a,b]表示a是b的先修课程
            int b=prereqs[i][1];
            adj[a].push_back(b);                //用<a,b>表示即先修a,再修b
        }
        for(int i=0;i<n;i++)
            bfs(n,i);
        vector<bool> ans(ques.size()); 
        for (int i=0;i<ques.size();i++)
            ans[i]=path[ques[i][0]][ques[i][1]];
        return ans;
    }
    void bfs(int n,int i) {                                   //广度优先遍历算法
        queue<int> qu;                              //定义一个队列qu
        path[i][i]=true;                            //自己到自己有路径
        qu.push(i);                                 //课程i进队
        while (!qu.empty()) {                       //队不空循环
            int cnt=qu.size();
            for(int j=0;j<cnt;j++) {
                int u=qu.front(); qu.pop();         //出队课程u
                for(int v:adj[u]) {                 //考虑课程u的出边课程v
                    if(path[i][v]) continue;
                    path[i][v]=true;                //表示课程i到v有路径
                    qu.push(v);                     //课程v进队
                }
            }
        }
    }
};

File: vol2/ch18/LeetCode1462-2.py
class Solution:
    def checkIfPrerequisite(self,n:int,prereqs,ques) -> List[bool]:
        self.adj=[[] for _ in range(0,n)]          #邻接表
        self.path=[[False for _ in range(n)] for _ in range(n)];
        for i in range(0,len(prereqs)):    #由列表创建邻接表
            a=prereqs[i][0]                #[a,b]表示a是b的先修课程
            b=prereqs[i][1]
            self.adj[a].append(b)                #用<a,b>表示即先修a,再修b
        for i in range(0,n):
            self.bfs(n,i)
        ans=[False for _ in range(0,len(ques))]     #存放答案
        for i in range(0,len(ques)):
            ans[i]=self.path[ques[i][0]][ques[i][1]]
        return ans
    def bfs(self,n,i):                              #从课程i出发广度优先搜索
        qu=deque()                              #定义一个队列qu
        self.path[i][i]=True                            #自己到自己有路径
        qu.append(i)                                 #课程i进队
        while qu:                      #队不空循环
            cnt=len(qu)
            for j in range(0,cnt):
                u=qu.popleft()         #出队课程u
                for v in self.adj[u]:                 #考虑课程u的出边课程v
                    if self.path[i][v]:continue;
                    self.path[i][v]=True                #表示课程i到v有路径
                    qu.append(v)                     #课程v进队
